python/ray/data/datasource/huggingface_datasource.py

```diff
@@ -88,11 +88,35 @@ def get_read_tasks(
         # Note: `parallelism` arg is currently not used by HuggingFaceDatasource.
         # We always generate a single ReadTask to perform the read.
         _check_pyarrow_version()
+        import numpy as np
+        import pandas as pd
         import pyarrow
 
         def _read_dataset(dataset: "datasets.IterableDataset") -> Iterable[Block]:
             for batch in dataset.with_format("arrow").iter(batch_size=self._batch_size):
-                block = pyarrow.Table.from_pydict(batch)
+                # HuggingFace IterableDatasets do not fully support methods like
+                # `set_format`, `with_format`, and `formatted_as`, so the dataset
+                # can return whatever is the default configured batch type, even if
+                # the format is manually overriden before iterating above.
+                # Therefore, we limit support to batch formats which have native
+                # block types in Ray Data (pyarrow.Table, pd.DataFrame),
+                # or can easily be converted to such (dict, np.array).
+                # See: https://github.com/huggingface/datasets/issues/3444
+                if not isinstance(batch, (pyarrow.Table, pd.DataFrame, dict, np.array)):
+                    raise ValueError(
+                        f"Batch format {type(batch)} is not supported. Only the "
+                        f"following batch formats are supported: "
+                        f"dict (corresponds to `None` in `dataset.with_format()`), "
+                        f"pyarrow.Table, np.array, pd.DataFrame."
+                    )
+                # With Ray Data strict mode, np.arrays must be wrapped in a dict
+                # (subsequently converted to pyarrow.Table).
```

Comment:
Would avoid referencing "strict mode" here. It's meaningless unless you have context from the time we made this change